### PR TITLE
Physical_oM: Creation of BulkSolids

### LINKED ISSUE #1017: Physical_oM: Creation of a BulkMaterial object
#### Description:

As discussed with @michaelhoehn and @ajensen19 - the concept of a bulk material object is useful in the context of the https://github.com/BHoM/LifeCycleAssessment_Toolkit 

This object will be defined by something like

``` C# 
public class BulkMaterial : BHoMObject, IElementM
{
    public virtual List<BH.oM.Geometry.ISolid> Geometry { get; set; }
    public virtual MaterialComposition MaterialComposition { get; set; }

}

```

### PR DESCRIPTION
**Pull request by michaelhoehn**
### Issues addressed by this PR

Closes #1017 

### Test files

[Test File](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Test%20Scripts/BHoM/Physical_oM-Issue1017-BulkMaterials/201029_BulkMaterials.gh?csf=1&web=1&e=CXPg3d)

### Changelog

- Creation of `ISolid`
- Creation of `BulkSolids`
- Creation of `VolumeFragment`

### CODE REVIEW

**Comment by michaelhoehn** [comment]
DNM placed until discussion and/or the resolution of IElement3D for solids.

**Review by IsakNaslundBh — COMMENTED** [review]
Small comment on the material composition on the interface in particular

File: Physical_oM/Elements/ISolid.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+namespace BH.oM.Physical.Elements
+{
+    public interface ISolid : IBHoMObject, IPhysical, IElementM //, IElement3D
+    {
+        /***************************************************/
+        /**** Properties                                ****/
+        /***************************************************/
+        List<BH.oM.Geometry.ISolid> Geometry { get; set; }
+        MaterialComposition MaterialComposition { get; set; }
```
[the remark this answers is not in the document]
Comment (michaelhoehn) [reply Physical_oM/Elements/ISolid.cs:36]: Agreed! Thanks @IsakNaslundBh

**Comment by pawelbaran** [comment]
Pardon me coming from a slightly different angle, but following [my comment here](https://github.com/BHoM/BHoM/issues/1017#issuecomment-707109986), does the proposed implementation allow the cases with unknown/unsupported geometry, but known volume? I think such functionality can be powerful in two cases:
- lacking convert (imagine a blob that cannot be converted to BHoM due to its complexity, but can be queried for volume in the software it comes from - prime example being Revit_Toolkit with its lack of support for solids)
- elements of high geometrical complexity (say, facade mullion with a funky shape - the user may not want to drag the geometry around when only the volume as a number is relevant)

**Comment by michaelhoehn** [comment]
> * lacking convert (imagine a blob that cannot be converted to BHoM due to its complexity, but can be queried for volume in the software it comes from - prime example being Revit_Toolkit with its lack of support for solids)
> * elements of high geometrical complexity (say, facade mullion with a funky shape - the user may not want to drag the geometry around when only the volume as a number is relevant)

@pawelbaran good suggestions here. We could solve this in a number of different ways. This current implementation provides a strongly typed object that is capable of calculating the object's volume. Alternatively, we could create another object that doesn't provide this functionality but simply has a Volume property. We could also make use of fragments by simply allowing the user to append this data to the object if needed. 

Any other thoughts on how we could accomplish this? May be best to raise subsequent issues to catch the additional functionality. 

FYI: @al-fisher @FraserGreenroyd @IsakNaslundBh

**Comment by michaelhoehn** [comment]
@BHoMBot check installer

**Comment by pawelbaran** [comment]
I am not sure if I like creating an object with volume property but no geometry. However, a `VolumeFragment` could be a nice idea - `Query.Volume(BulkMaterial)` could query the geometry first, and then the fragment if the former fails. How do you like that @al-fisher @michaelhoehn? This could unlock the stale PR on Revit_Toolkit that got parked due to the issues with representation of volume without geometry (among others): https://github.com/BHoM/Revit_Toolkit/pull/774. @enarhi FYI

**Comment by al-fisher** [comment]
Yes - agreed - the volume fragment sounds quite elegant

**Comment by michaelhoehn** [comment]
@pawelbaran @al-fisher I've had a go at creating a VolumeFragment and have edited the `BH.Engine.Physical.Query.SolidVolume` method to check for VolumeFragment values if no geometry specific query values are reported. It will also provide a note to the user if a fragment value is being used.

You will find some initial scenarios in the linked test script, although I don't have a particular test case for this functionality, it sounds like it could be useful elsewhere.

**Comment by IsakNaslundBh** [comment]
> I am not sure if I like creating an object with volume property but no geometry.

I do not really have anything against this actually. Almost would prefer this over a `VolumeFragment`, as the latter have a big chance of confusion as to what volume is being used, and where.

Just having a sort of fallback object that simply has volume and material(composition) sounds both simpler and safer to me.

Might be missing some usecase here though?

**Comment by al-fisher** [comment]
@michaelhoehn - picking up on @IsakNaslundBh and @pawelbaran's comments on the two options for a generic (non-solid geometry based) bulk object - worth a bit more discussion probably
Can see Fragment working at most generic level - but can also see point made above over the possible confusion.

I would therefore raise a new issue for this to continue planning and take out of the scope of your PR here as not needed for your BulkSolids

**Comment by michaelhoehn** [comment]
> I do not really have anything against this actually. Almost would prefer this over a `VolumeFragment`, as the latter have a big chance of confusion as to what volume is being used, and where.
> 
> Just having a sort of fallback object that simply has volume and material(composition) sounds both simpler and safer to me.

Agreed. 

> I would therefore raise a new issue for this to continue planning and take out of the scope of your PR here as not needed for your BulkSolids

Agreed, will be removing `VolumeFragment` in the next commit unless a more concrete use case can be made here. Following the removal, I will raise another issue to discuss the implementation of a secondary "bulk" object that contains a materialComposition and a Volume property.

**Comment by pawelbaran** [comment]
> I do not really have anything against this actually. Almost would prefer this over a VolumeFragment, as the latter have a big chance of confusion as to what volume is being used, and where.

That could be OK too, but only if both would implement a common interface, to make sure they can be queried easily on case of mixed collections.

@michaelhoehn, I believe https://github.com/BHoM/Revit_Toolkit/pull/774 is a very concrete use case (once we have the null geometry `BulkMaterial` sorted, Revit_Toolkit will be able to return a very precise material takeoff for _any_ object in Revit). But happy with resolving it in a separate PR, if that will make things easier.

**Comment by michaelhoehn** [comment]
@BHoMBot check installer

**Comment by michaelhoehn** [comment]
@BHoMBot check installer

**Review by al-fisher — APPROVED** [review]
LGTM - happy to approve along with Engine PR

**Pull request merged**